### ISSUE TITLE
[NetLinkSocket Error 7]: Socket::initSocket: error in socket connection/bind"

### ISSUE THREAD

**Issue by samaronesalles** [JacobFischer/netlinkwrapper#16]
Please any help?

line with error: 
const client = new SocketClientTCP(configTCP.port, configTCP.host);

the exception is:
[NetLinkSocket Error 7]: Socket::initSocket: error in socket connection/bind"

**Comment by JacobFischer** [comment]
Do you have a TCP Server listening to connect to? Without one the constructor will throw that error you are experiencing.

Does the example from the README run for you?:

```js
const { SocketClientTCP, SocketServerTCP } = require('netlinkwrapper');

const port = 33333;
const server = new SocketServerTCP(port);
const client = new SocketClientTCP(port, 'localhost');

const serverSends = 'hello world!';
const serversClient = server.accept();
serversClient.send(serverSends);

const clientReceived = client.receive();

const identical = serverSends === clientReceived.toString(); // should be true

console.log('the two strings are identical?', identical);

client.disconnect();
server.disconnect();
```

It should run successfully. If you omit the `const server = new SocketServerTCP(port);` line you'll see the `new SocketClientTCP` throws with the same error you are seeing for presumably the same reason, because a Client TCP socket requires a listening server to bind to.

**Comment by samaronesalles** [comment]
> Do you have a TCP Server listening to connect to? Without one the constructor will throw that error you are experiencing.
> 
> Does the example from the README run for you?:
> 
> ```js
> const { SocketClientTCP, SocketServerTCP } = require('netlinkwrapper');
> 
> const port = 33333;
> const server = new SocketServerTCP(port);
> const client = new SocketClientTCP(port, 'localhost');
> 
> const serverSends = 'hello world!';
> const serversClient = server.accept();
> serversClient.send(serverSends);
> 
> const clientReceived = client.receive();
> 
> const identical = serverSends === clientReceived.toString(); // should be true
> 
> console.log('the two strings are identical?', identical);
> 
> client.disconnect();
> server.disconnect();
> ```
> 
> It should run successfully. If you omit the `const server = new SocketServerTCP(port);` line you'll see the `new SocketClientTCP` throws with the same error you are seeing for presumably the same reason, because a Client TCP socket requires a listening server to bind to.

Yes, i have this. look:
------------------------------

`
const blaster = require('../functions/blaster');
const delay = require('delay');                  // Documentaçaõ: https://www.npmjs.com/package/delay

const configTCP = require('../config/configTCP');
const { SocketClientTCP } = require('netlinkwrapper');

module.exports = {

    async enviar_fun(action, codEmp, cnpj, mensagem) {
        try {

            const client = new SocketClientTCP(configTCP.port, configTCP.host);
            let answer = "";

            let msg = '';
            msg += 'CONEXAO_FINALIDADE=API\n';
            msg += 'CLIENT_TIPO=WEB\n';
            msg += 'CLIENT_VERSAO=1.01\n';
            msg += `CLIENT_IDCOPIA=${codEmp}\n`;
            msg += `CLIENT_CNPJ=${cnpj}`;

            msg = blaster.input_slice(msg);

            // Conectando...
            await client.send(msg);
            answer = await client.receive().toString();

            answer = blaster.removeBytes(answer);
            if (answer != "OK")                               // Tentar simular situacao para testar
                throw new Error(`Conexão não passou na validação: ${answer}`);

            // Enviando Action...
            msg = blaster.input_slice(action);

            await client.send(msg);

            answer = await client.receive().toString();

            answer = blaster.removeBytes(answer);
            if (answer != "START")                            // Tentar simular situacao para testar
                throw new Error(`START esperado, mas encontrado: ${answer}`);


            // Verificar se servidor está pronto para receber a mensagem...
            answer = await client.receive().toString();

            answer = blaster.removeBytes(answer);
            if (answer != "PROCESSANDO")                      // Tentar simular situacao para testar
                throw new Error(`PROCESSANDO esperado, mas encontrado: ${answer}`);



            // Enviando Mensagem...
            msg = blaster.input_slice(mensagem);
            await client.send(msg);

            answer = await client.receive().toString();

            answer = blaster.removeBytes(answer);

            if (answer.indexOf("TIMEOUT") >= 0)
                throw new Error(`Servidor excedeu o tempo limite para resposta. ${answer}`);

            client.disconnect();

            return answer;

        } catch (error) {
            // if (client != undefined)
            //     client.disconnect();

            throw error;
        }
    },

};
`

**Comment by JacobFischer** [comment]
I suggest trying to use node's [net module](https://nodejs.org/api/net.html) instead. When it cannot connect it will give you richer error messages, such as:

```
Error: connect ECONNREFUSED 127.0.0.1:33334
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1144:16) {
  errno: -4078,
  code: 'ECONNREFUSED',
  syscall: 'connect',
  address: '127.0.0.1',
  port: 33334
}
```

Node's error messages are much clearer when the actual error is the connection is being refused.

I will try to update this module with richer error messages akin to net to help developers debug their program. Currently I just essentially re-throw the underlying socket library's errors from C++ to JS.

As a side note. In your example code you are running it in an `async` context. Because you can/are running async code I'm not sure why this module is being used. `netlinkwrapper`'s main appeal is **synchronous** usage. It is capable of blocking the event loop to do network communication, which to my knowledge is still novel to this module. However because you are running in an async context already you can use the event loop and use proper node-style callbacks. This is what I would recommend.

For example, when you do:
```js
await client.send(msg);
```

`SocketClientTCP` never returns promises, so the `await` keyword does nothing here. 
Instead if you where using a `net.Socket` you could do proper async usage:

```js
await new Promise((resolve, reject) => clientNetSocket.write(msg, (err) => {
  if (err) {
    reject(err);
  } else {
    resolve();
  }
});
```

**Comment by samaronesalles** [comment]
I understood,

I'm sorry for the bad code, I'm a beginner in nodejs :)

I will remove all "await" in the calls to be correct ..

My need is to work synchronously, before I installed its component, I was using the "net module", but I couldn't make it work synchronously. Using the "net module" this error did not happen.

It was then, that I found its component, and I was interested, because it does for me, which I couldn't do using the "net module" due to my lack of knowledge.

Yesterday I had difficulties in the installation, but after being able to install it, I did tests and it was working. But today you are giving this error without changing the code.

** Sorry for my English, I'm Brazilian and I'm using the google translator.

**Comment by samaronesalles** [comment]
problem solved. Thanks a lot

Before :
const client = new SocketClientTCP(configTCP.port, configTCP.host);

After:
const client = new SocketClientTCP(6000, 'localhost');

**Comment by JacobFischer** [comment]
It is no problem. I want to help and ideally discover and fix bugs when they pop up.

Your code is not bad, I was just pointing out correct usage of this module. If you are just learning Node.js then the example you provided shows you are doing well thus far in my opinion.

Also I presumed Portuguese was your primary language seeing as the code you posted had Portuguese comments. Thank you for not being too intimidated to reach out in English for me.

I'm glad you resolved your issue. Though if possible could you tell me what values `configTCP` had? the `SocketClientTCP` constructor should throw on invalid port numbers and host strings before ever attempting a connection. If it is not that too is a bug I need to fix.

**Comment by JacobFischer** [comment]
I just ran a quick test and this reproduced your issue:

```js
const client = new SocketClientTCP(undefined, undefined);
```

That should not be allowed. So this is indeed a bug. I will re-open this issue and better inspect arguments to throw a TypeError when the arguments are not the correct types, before it ever tries to connect.

**Comment by samaronesalles** [comment]
Thank you,

It will be a pleasure if I can contribute to improving its component. This component helped me to overcome my difficulty with the "net module", doing for me what I was unable to do.

Before:
```js
const client = new SocketClientTCP (configTCP.port, configTCP.host);
```
"configTCP" is a "* .js" file and has the code below:
```js
'use strict';

module.exports = {
    ms_timeOut: 100,
    port: process.env.TCPSERVER_PORT,
    host: process.env.TCPSERVER_HOST
}
```

"process.env.TCPSERVER_PORT" and "process.env.TCPSERVER_HOST" are in the ".env.testing" file and their contents are:
```js
# .env, this file can be commit to repo. use this as an example to create in production

# variables to initialization the server
APP_PORT = 3000

# variables to connection with TCP-LBCServer
TCPSERVER_HOST = localhost
TCPSERVER_PORT = 6000
```
------------------------------------

Important: After I tested the solution, there was a moment when the error came back, I closed vscode, and went to work on another company project that is Delphi ... After a few hours I saw your answer, and opened VScode again to test it, and the problem didn't happen. Is there a risk of it becoming an intermittent problem?

Once again I want to thank you for the attention you are giving me, and I want to use your component in this important project and what I can contribute with you to improve it, count on me.

**Comment by JacobFischer** [comment]
Thank you.

This should be resolved in the newly released [2.0.2 version](https://github.com/JacobFischer/netlinkwrapper/releases/tag/v2.0.2).

There may be some intermittency issues, however as is the nature of them they are difficult to track down. I will keep an eye out for connection binding issues in the future. For now this release should fix the issues with invalid parameters passed to the constructor.
